### PR TITLE
GCM: record limit counter gets reset on AAD changes

### PR DESCRIPTION
**Pull request by paulidale**
It shouldn't be.  This moves the reset to the init function instead.

- [ ] documentation is added or updated
- [ ] tests are added or updated

### CODE REVIEW

**Comment by slontis** [comment]
In the e_aes.c code the ctrl call to  EVP_CTRL_AEAD_TLS1_AAD set this to 0, so I think it is probably wrong in earlier OpenSSL versions also?

**Comment by paulidale** [comment]
Just 3.x and later.

**Comment by slontis** [comment]
Doesnt that still mean legacy paths using engines could be broken in 3.X?

**Review by hlandau — APPROVED** [review]

File: providers/implementations/ciphers/ciphercommon_gcm.c [hunk 1]
```diff
@@ -48,6 +55,7 @@ static int gcm_init(void *vctx, const unsigned char *key, size_t keylen,
         return 0;
 
     ctx->enc = enc;
+    ctx->tls_enc_records = 0;
```
Comment (t8m) [inline providers/implementations/ciphers/ciphercommon_gcm.c:58]: Should it be moved to the `if (key != NULL)` block? IMO the counter should be reset only on key changes.

Comment (paulidale) [reply providers/implementations/ciphers/ciphercommon_gcm.c:58]: Yes :(

**Review by hlandau — APPROVED** [review]

**Review by t8m — APPROVED** [review]

**Comment by t8m** [comment]
Merged to master and 3.0 branches. Thank you.

**Pull request closed without merging**